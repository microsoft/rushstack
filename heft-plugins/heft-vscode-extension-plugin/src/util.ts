// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { Executable, IExecutableSpawnOptions, IWaitForExitResult } from '@rushstack/node-core-library';
import type { ChildProcess } from 'node:child_process';
import * as path from 'node:path';
import { TerminalStreamWritable, TerminalProviderSeverity, ITerminal } from '@rushstack/terminal';

export async function executeAndWaitAsync(
  terminal: ITerminal,
  command: string,
  args: string[],
  options: Omit<IExecutableSpawnOptions, 'stdio'> = {}
): Promise<IWaitForExitResult<string>> {
  const childProcess: ChildProcess = Executable.spawn(command, args, {
    ...options,
    stdio: [
      'ignore', // stdin
      'pipe', // stdout
      'pipe' // stderr
    ]
  });
  childProcess.stdout?.pipe(
    new TerminalStreamWritable({
      terminal,
      severity: TerminalProviderSeverity.log
    })
  );
  childProcess.stderr?.pipe(
    new TerminalStreamWritable({
      terminal,
      severity: TerminalProviderSeverity.error
    })
  );
  const result: IWaitForExitResult<string> = await Executable.waitForExitAsync(childProcess, {
    encoding: 'utf8'
  });
  return result;
}

const vsceBasePackagePath: string = require.resolve('@vscode/vsce/package.json');
export const vsceScriptPath: string = path.resolve(vsceBasePackagePath, '../vsce');
