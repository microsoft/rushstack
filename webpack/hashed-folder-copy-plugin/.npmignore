# Ignore everything by default
**

# Use negative patterns to bring back the specific things we want to publish
!/bin/**
!/lib/**
!/dist/**
!ThirdPartyNotice.txt
!/EULA/**

# Ignore certain files in the above folder
/dist/*.stats.*
/lib/**/test/**
/lib/**/*.js.map
/dist/**/*.js.map

# NOTE: These don't need to be specified, because NPM includes them automatically.
#
# package.json
# README (and its variants)
# CHANGELOG (and its variants)
# LICENSE / LICENCE

## Project specific definitions
# -----------------------------
