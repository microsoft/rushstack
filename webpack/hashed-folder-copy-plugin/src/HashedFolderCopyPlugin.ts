// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { Async } from '@rushstack/node-core-library';
import type { CallExpression, Expression, UnaryExpression } from 'estree';
import type webpack from 'webpack';
import type glob from 'fast-glob';

import {
  type IHashedFolderDependency,
  getHashedFolderDependencyForWebpackInstance
} from './HashedFolderDependency';
import type { BasicEvaluatedExpression } from './webpackTypes';

interface IParserHelpers {
  evaluateToString: (type: string) => (exp: UnaryExpression) => BasicEvaluatedExpression;
  toConstantDependency: (parser: webpack.Parser, type: string) => (exp: Expression) => true;
}

// TODO: Use the compiler's webpack exports instead of requiring from webpack
const ParserHelpers: IParserHelpers = require('webpack/lib/javascript/JavascriptParserHelpers');

const PLUGIN_NAME: 'hashed-folder-copy-plugin' = 'hashed-folder-copy-plugin';

const EXPRESSION_NAME: 'requireFolder' = 'requireFolder';

interface IAcornNode<TExpression> {
  computed: boolean | undefined;
  elements: IAcornNode<unknown>[];
  key: IAcornNode<unknown> | undefined;
  name: string | undefined;
  properties: IAcornNode<unknown>[] | undefined;
  type: 'Literal' | 'ObjectExpression' | 'Identifier' | 'ArrayExpression' | unknown;
  value: TExpression;
}

export function renderError(errorMessage: string): string {
  return `(function () { throw new Error(${JSON.stringify(errorMessage)}); })()`;
}

/**
 * @public
 */
export class HashedFolderCopyPlugin implements webpack.WebpackPluginInstance {
  public apply(compiler: webpack.Compiler): void {
    const webpack: typeof import('webpack') = compiler.webpack;
    const { HashedFolderDependency, HashedFolderDependencyTemplate } =
      getHashedFolderDependencyForWebpackInstance(webpack);

    compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation: webpack.Compilation) => {
      compilation.dependencyTemplates.set(HashedFolderDependency, new HashedFolderDependencyTemplate());
    });

    const hashedFolderDependencies: IHashedFolderDependency[] = [];

    compiler.hooks.thisCompilation.tap(
      PLUGIN_NAME,
      (compilation: webpack.Compilation, { normalModuleFactory }) => {
        compilation.hooks.finishModules.tapPromise(PLUGIN_NAME, async () => {
          const { inputFileSystem } = compiler;

          const notImplementedFunction: () => never = () => {
            throw new Error('Not implemented');
          };
          const globFs: glob.FileSystemAdapter = {
            lstat: inputFileSystem?.lstat?.bind(inputFileSystem) ?? notImplementedFunction,
            stat: inputFileSystem?.stat?.bind(inputFileSystem) ?? notImplementedFunction,
            lstatSync: notImplementedFunction,
            statSync: notImplementedFunction,
            readdir: inputFileSystem?.readdir?.bind(inputFileSystem) ?? notImplementedFunction,
            readdirSync: notImplementedFunction
          } as unknown as glob.FileSystemAdapter; // The Webpack typings are wrong on `readdir`

          await Async.forEachAsync(
            hashedFolderDependencies,
            async (hashedFolderDependency) => {
              await hashedFolderDependency.processAssetsAsync(compilation, globFs);
            },
            { concurrency: 10 }
          );
        });

        const handler: (parser: webpack.javascript.JavascriptParser) => void = (
          parser: webpack.javascript.JavascriptParser
        ) => {
          parser.hooks.call.for(EXPRESSION_NAME).tap(PLUGIN_NAME, (baseExpression: Expression) => {
            const expression: CallExpression = baseExpression as CallExpression;

            let errorMessage: string | undefined;
            let requireFolderOptions: IRequireFolderOptions | undefined = undefined;
            if (expression.arguments.length !== 1) {
              errorMessage = `Exactly one argument is required to be passed to "${EXPRESSION_NAME}"`;
            } else {
              const argument: IAcornNode<IRequireFolderOptions> = expression
                .arguments[0] as IAcornNode<IRequireFolderOptions>;
              try {
                requireFolderOptions = this._evaluateAcornNode(argument) as IRequireFolderOptions;
              } catch (e) {
                errorMessage = (e as Error).message;
              }

              if (requireFolderOptions) {
                if (
                  !requireFolderOptions.outputFolder ||
                  typeof requireFolderOptions.outputFolder !== 'string'
                ) {
                  errorMessage = 'The options object must have a "outputFolder" property that is a string';
                } else if (!requireFolderOptions.sources || !Array.isArray(requireFolderOptions.sources)) {
                  errorMessage = 'The options object must have a "sources" property that is an array';
                } else {
                  for (const source of requireFolderOptions.sources) {
                    if (!source.globsBase || typeof source.globsBase !== 'string') {
                      errorMessage = 'Each "sources" element must have a string "globsBase" property';
                    } else if (
                      !source.globPatterns ||
                      !Array.isArray(source.globPatterns) ||
                      source.globPatterns.some(
                        (globPattern) => !globPattern || typeof globPattern !== 'string'
                      )
                    ) {
                      errorMessage =
                        'Each "sources" element must have a "globPatterns" property that is an array of glob strings';
                    }
                  }
                }
              }
            }

            const currentModule: webpack.NormalModule = parser.state.current;
            let dependency: webpack.dependencies.NullDependency;
            if (!requireFolderOptions) {
              const errorText: string = renderError(errorMessage!);
              dependency = new webpack.dependencies.ConstDependency(errorText, expression.range!);
              if (expression.loc) {
                dependency.loc = expression.loc;
              }

              compilation.errors.push(new webpack.WebpackError(errorMessage));
            } else {
              const hashedFolderDependency: IHashedFolderDependency = new HashedFolderDependency(
                requireFolderOptions,
                expression.range!,
                expression.loc
              );
              hashedFolderDependencies.push(hashedFolderDependency);
              dependency = hashedFolderDependency;
            }

            currentModule.addDependency(dependency);
          });

          parser.hooks.evaluateTypeof
            .for(EXPRESSION_NAME)
            .tap(PLUGIN_NAME, ParserHelpers.evaluateToString('function'));

          parser.hooks.typeof
            .for(EXPRESSION_NAME)
            .tap(PLUGIN_NAME, ParserHelpers.toConstantDependency(parser, JSON.stringify('function')));
        };

        normalModuleFactory.hooks.parser.for('javascript/auto').tap(PLUGIN_NAME, handler);
        normalModuleFactory.hooks.parser.for('javascript/dynamic').tap(PLUGIN_NAME, handler);
      }
    );
  }

  private _evaluateAcornNode(node: IAcornNode<unknown>): unknown {
    switch (node.type) {
      case 'Literal': {
        return node.value;
      }

      case 'ObjectExpression': {
        const result: Record<string, unknown> = {};

        for (const property of node.properties!) {
          const keyNode: IAcornNode<unknown> = property.key!;
          if (keyNode.type !== 'Identifier' || keyNode.computed) {
            throw new Error('Property keys must be non-computed identifiers');
          }

          const key: string = keyNode.name!;
          const value: unknown = this._evaluateAcornNode(property.value as IAcornNode<unknown>);
          result[key] = value;
        }

        return result;
      }

      case 'ArrayExpression': {
        return node.elements.map((element) => this._evaluateAcornNode(element));
      }

      default: {
        throw new Error(`Unsupported node type: "${node.type}"`);
      }
    }
  }
}
