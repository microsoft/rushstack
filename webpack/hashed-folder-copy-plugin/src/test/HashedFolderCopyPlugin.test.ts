// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

jest.mock('node:path', () => {
  const path: typeof import('path') = jest.requireActual('path');
  return path.posix;
});

jest.mock(
  'fast-glob/out/providers/provider',
  () => {
    const path: typeof import('path') = jest.requireActual('path');
    const { default: provider } = jest.requireActual('fast-glob/out/providers/provider');
    provider.prototype._getRootDirectory = function (task: { base: string }) {
      // fast-glob calls `path.resolve` which doesn't work correctly with the MemFS volume while running on Windows
      return path.posix.resolve(this._settings.cwd, task.base);
    };
    return { default: provider };
  },
  {}
);

import type { default as webpack, Stats, InputFileSystem, OutputFileSystem } from 'webpack';
import type { Volume } from 'memfs/lib/volume';
import type { FileSystem, FolderItem } from '@rushstack/node-core-library';

jest.setTimeout(1e9);

interface IFilePaths {
  itemRelativePath: string;
  itemAbsolutePath: string;
}

async function* enumerateFilesAsync(
  fileSystem: typeof FileSystem,
  absolutePath: string,
  relativePath: string
): AsyncIterable<IFilePaths> {
  const folderItems: FolderItem[] = await fileSystem.readFolderItemsAsync(absolutePath);
  for (const item of folderItems) {
    const name: string = item.name;
    const itemRelativePath: string = `${relativePath}/${name}`;
    const itemAbsolutePath: string = `${absolutePath}/${name}`;
    if (item.isDirectory()) {
      yield* enumerateFilesAsync(fileSystem, itemAbsolutePath, itemRelativePath);
    } else if (item.isFile()) {
      yield { itemRelativePath, itemAbsolutePath };
    } else {
      throw new Error(`Unexpected item type`);
    }
  }
}

async function readFolderAsync(
  absolutePath: string,
  relativePath: string,
  outputObject: Record<string, string>
): Promise<void> {
  const { Async, FileSystem } = await import('@rushstack/node-core-library');

  const files: AsyncIterable<IFilePaths> = enumerateFilesAsync(FileSystem, absolutePath, relativePath);
  await Async.forEachAsync(
    files,
    async ({ itemRelativePath, itemAbsolutePath }) => {
      const fileContents: string = await FileSystem.readFileAsync(itemAbsolutePath);
      outputObject[itemRelativePath] = fileContents;
    },
    { concurrency: 50 }
  );
}

async function runTestAsync(inputFolderPath: string): Promise<void> {
  const [{ Volume }, { default: webpack }, { promisify }, { HashedFolderCopyPlugin }] = await Promise.all([
    import('memfs/lib/volume'),
    import('webpack'),
    import('util'),
    import('../HashedFolderCopyPlugin')
  ]);

  const memoryFileSystem: Volume = new Volume();
  const folderContents: Record<string, string> = {};
  await readFolderAsync(inputFolderPath, '', folderContents);
  memoryFileSystem.fromJSON(folderContents, '/src');

  const compiler: webpack.Compiler = webpack({
    entry: {
      main: '/entry.js'
    },
    output: {
      path: '/release',
      filename: '[name].js'
    },
    optimization: {
      minimizer: []
    },
    context: '/',
    mode: 'production',
    plugins: [new HashedFolderCopyPlugin()]
  });

  compiler.inputFileSystem = memoryFileSystem as unknown as InputFileSystem;
  compiler.outputFileSystem = memoryFileSystem as unknown as OutputFileSystem;

  const stats: Stats | undefined = await promisify(compiler.run.bind(compiler))();
  await promisify(compiler.close.bind(compiler));
  if (!stats) {
    throw new Error(`Expected stats`);
  }
  const { errors, warnings } = stats.toJson('errors-warnings');
  expect(errors).toMatchSnapshot('Errors');
  expect(warnings).toMatchSnapshot('Warnings');

  const results: {} = memoryFileSystem.toJSON('/release');
  expect(results).toMatchSnapshot('Content');
}

describe('HashedFolderCopyPlugin', () => {
  it('Handles the null case', async () => {
    await runTestAsync(`${__dirname}/scenarios/nullCase`);
  });

  it('Handles globbing a local folder', async () => {
    await runTestAsync(`${__dirname}/scenarios/localFolder`);
  });

  it('Handles globbing a package reference', async () => {
    await runTestAsync(`${__dirname}/scenarios/packageReference`);
  });
});
