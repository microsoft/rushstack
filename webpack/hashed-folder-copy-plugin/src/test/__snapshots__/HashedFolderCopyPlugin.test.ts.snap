// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HashedFolderCopyPlugin Handles globbing a local folder: Content 1`] = `
Object {
  "/release/assets_1a20fc5e5390d91fee246bf5dbf1300e/a.txt": "local-folder-AAA",
  "/release/assets_1a20fc5e5390d91fee246bf5dbf1300e/b.txt": "local-folder-BBB",
  "/release/assets_1a20fc5e5390d91fee246bf5dbf1300e/subfolder/c.txt": "local-folder-CCC",
  "/release/main.js": "/******/ (() => { // webpackBootstrap
/******/ 	\\"use strict\\";
/******/ 	// The require scope
/******/ 	var __webpack_require__ = {};
/******/ 	
/************************************************************************/
/******/ 	/* webpack/runtime/global */
/******/ 	(() => {
/******/ 		__webpack_require__.g = (function() {
/******/ 			if (typeof globalThis === 'object') return globalThis;
/******/ 			try {
/******/ 				return this || new Function('return this')();
/******/ 			} catch (e) {
/******/ 				if (typeof window === 'object') return window;
/******/ 			}
/******/ 		})();
/******/ 	})();
/******/ 	
/******/ 	/* webpack/runtime/publicPath */
/******/ 	(() => {
/******/ 		var scriptUrl;
/******/ 		if (__webpack_require__.g.importScripts) scriptUrl = __webpack_require__.g.location + \\"\\";
/******/ 		var document = __webpack_require__.g.document;
/******/ 		if (!scriptUrl && document) {
/******/ 			if (document.currentScript && document.currentScript.tagName.toUpperCase() === 'SCRIPT')
/******/ 				scriptUrl = document.currentScript.src;
/******/ 			if (!scriptUrl) {
/******/ 				var scripts = document.getElementsByTagName(\\"script\\");
/******/ 				if(scripts.length) {
/******/ 					var i = scripts.length - 1;
/******/ 					while (i > -1 && (!scriptUrl || !/^http(s?):/.test(scriptUrl))) scriptUrl = scripts[i--].src;
/******/ 				}
/******/ 			}
/******/ 		}
/******/ 		// When supporting browsers where an automatic publicPath is not supported you must specify an output.publicPath manually via configuration
/******/ 		// or pass an empty string (\\"\\") and set the __webpack_public_path__ variable from your code to use your own logic.
/******/ 		if (!scriptUrl) throw new Error(\\"Automatic publicPath is not supported in this browser\\");
/******/ 		scriptUrl = scriptUrl.replace(/#.*$/, \\"\\").replace(/\\\\?.*$/, \\"\\").replace(/\\\\/[^\\\\/]+$/, \\"/\\");
/******/ 		__webpack_require__.p = scriptUrl;
/******/ 	})();
/******/ 	
/************************************************************************/

// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
// eslint-disable-next-line no-console
const path = __webpack_require__.p + \\"assets_1a20fc5e5390d91fee246bf5dbf1300e/\\";
// eslint-disable-next-line no-console
console.log(path);
// eslint-disable-next-line no-console
console.log(\\"function\\");
//# sourceMappingURL=entry.js.map
/******/ })()
;",
}
`;

exports[`HashedFolderCopyPlugin Handles globbing a local folder: Errors 1`] = `Array []`;

exports[`HashedFolderCopyPlugin Handles globbing a local folder: Warnings 1`] = `Array []`;

exports[`HashedFolderCopyPlugin Handles globbing a package reference: Content 1`] = `
Object {
  "/release/assets_084b4ae6ce51eb9a614a9d1981afc6b7/a.txt": "unscoped-package-AAA",
  "/release/assets_084b4ae6ce51eb9a614a9d1981afc6b7/b.txt": "unscoped-package-BBB",
  "/release/assets_084b4ae6ce51eb9a614a9d1981afc6b7/subfolder/c.txt": "unscoped-package-CCC",
  "/release/assets_f2eede62be13dd0052e19c66d82f6cdb/a.txt": "scoped-package-AAA",
  "/release/assets_f2eede62be13dd0052e19c66d82f6cdb/b.txt": "scoped-package-BBB",
  "/release/assets_f2eede62be13dd0052e19c66d82f6cdb/subfolder/c.txt": "scoped-package-CCC",
  "/release/main.js": "/******/ (() => { // webpackBootstrap
/******/ 	\\"use strict\\";
/******/ 	// The require scope
/******/ 	var __webpack_require__ = {};
/******/ 	
/************************************************************************/
/******/ 	/* webpack/runtime/global */
/******/ 	(() => {
/******/ 		__webpack_require__.g = (function() {
/******/ 			if (typeof globalThis === 'object') return globalThis;
/******/ 			try {
/******/ 				return this || new Function('return this')();
/******/ 			} catch (e) {
/******/ 				if (typeof window === 'object') return window;
/******/ 			}
/******/ 		})();
/******/ 	})();
/******/ 	
/******/ 	/* webpack/runtime/publicPath */
/******/ 	(() => {
/******/ 		var scriptUrl;
/******/ 		if (__webpack_require__.g.importScripts) scriptUrl = __webpack_require__.g.location + \\"\\";
/******/ 		var document = __webpack_require__.g.document;
/******/ 		if (!scriptUrl && document) {
/******/ 			if (document.currentScript && document.currentScript.tagName.toUpperCase() === 'SCRIPT')
/******/ 				scriptUrl = document.currentScript.src;
/******/ 			if (!scriptUrl) {
/******/ 				var scripts = document.getElementsByTagName(\\"script\\");
/******/ 				if(scripts.length) {
/******/ 					var i = scripts.length - 1;
/******/ 					while (i > -1 && (!scriptUrl || !/^http(s?):/.test(scriptUrl))) scriptUrl = scripts[i--].src;
/******/ 				}
/******/ 			}
/******/ 		}
/******/ 		// When supporting browsers where an automatic publicPath is not supported you must specify an output.publicPath manually via configuration
/******/ 		// or pass an empty string (\\"\\") and set the __webpack_public_path__ variable from your code to use your own logic.
/******/ 		if (!scriptUrl) throw new Error(\\"Automatic publicPath is not supported in this browser\\");
/******/ 		scriptUrl = scriptUrl.replace(/#.*$/, \\"\\").replace(/\\\\?.*$/, \\"\\").replace(/\\\\/[^\\\\/]+$/, \\"/\\");
/******/ 		__webpack_require__.p = scriptUrl;
/******/ 	})();
/******/ 	
/************************************************************************/

// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
const scopedPackagePath = __webpack_require__.p + \\"assets_f2eede62be13dd0052e19c66d82f6cdb/\\";
const unscopedPackagePath = __webpack_require__.p + \\"assets_084b4ae6ce51eb9a614a9d1981afc6b7/\\";
// eslint-disable-next-line no-console
console.log(scopedPackagePath);
// eslint-disable-next-line no-console
console.log(unscopedPackagePath);
//# sourceMappingURL=entry.js.map
/******/ })()
;",
}
`;

exports[`HashedFolderCopyPlugin Handles globbing a package reference: Errors 1`] = `Array []`;

exports[`HashedFolderCopyPlugin Handles globbing a package reference: Warnings 1`] = `Array []`;

exports[`HashedFolderCopyPlugin Handles the null case: Content 1`] = `
Object {
  "/release/main.js": "/******/ (() => { // webpackBootstrap
/******/ 	\\"use strict\\";

// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
// eslint-disable-next-line no-console
console.log('test');
//# sourceMappingURL=entry.js.map
/******/ })()
;",
}
`;

exports[`HashedFolderCopyPlugin Handles the null case: Errors 1`] = `Array []`;

exports[`HashedFolderCopyPlugin Handles the null case: Warnings 1`] = `Array []`;
